File: examples/pyjevsim-time-advance/regulator_main.py
"""Regulator federate subprocess.

Cycles:
  1. Open RtiConnection to rtid, join the federation under --name.
  2. enable_time_regulation(--lookahead).
  3. (optional) enable_time_constrained.
  4. For --cycles iterations:
     - issue next_message_request(i * tick-step)
     - wait for *full* TimeAdvanceGrant on events() — accumulating
       any forced (partial) grants per IEEE 1516.1
  5. Write result JSON: name, lookahead, list of grant times.

M22 W3: the M21-era retry-on-TimeAdvancingState backoff loop is
gone. The symptom it worked around was an SDK-side semantics gap,
not a server race: forced grants (clearPending=false in
advance.go::decideGrant) leave the federate in time-advancing
state, and the federate must keep waiting on the same NER until
a full grant arrives. ``wait_for_full_grant`` accumulates forced
grants and returns only on the full grant.
"""

from __future__ import annotations

import asyncio
import sys
from typing import Any

# ruff: noqa: E402  (sys.path tweaks via _federate_common must precede imports)
from _federate_common import (  # type: ignore[import-not-found]
    common_parser,
    federation_spec,
    wait_for_full_grant,
    write_result,
)
from rti1516e.connection import RtiConnection  # type: ignore[import-not-found]


class GrantDrivenModel:
    """Minimal coupled model whose DEVS cycle is gated by real grants."""

    def __init__(self, tick_step: float) -> None:
        self.tick_step = tick_step
        self.logical_time = 0.0
        self.requested: list[float] = []
        self.grants: list[float] = []
        self.output_calls = 0
        self.completed_cycles = 0

    def time_advance(self) -> float:
        return self.tick_step

    def output_handler(self) -> dict[str, Any]:
        self.output_calls += 1
        return {}

    def internal_transition(self) -> None:
        self.completed_cycles += 1

    def external_transition(self, port: str, payload: Any) -> None:
        del port, payload

    def next_request(self) -> float:
        target = (self.completed_cycles + 1) * self.time_advance()
        self.requested.append(target)
        return target

    def observe_grant(self, granted: float) -> None:
        self.logical_time = granted
        self.grants.append(granted)


async def run(
    *,
    url: str,
    name: str,
    lookahead: float,
    cycles: int,
    tick_step: float,
    constrained: bool,
    grant_timeout: float,
) -> dict[str, Any]:
    print(f"regulator_main[{name}]: connecting to {url}", file=sys.stderr, flush=True)
    spec = federation_spec()
    model = GrantDrivenModel(tick_step)
    async with (
        RtiConnection.connect(url) as rti,
        rti.join_federation(spec, federate_name=name) as fed,
    ):
        print(f"regulator_main[{name}]: joined as handle {fed.handle}", file=sys.stderr, flush=True)

        await fed.enable_time_regulation(lookahead)
        if constrained:
            await fed.enable_time_constrained()

        for _ in range(cycles):
            t = model.next_request()
            await fed.next_message_request(t)
            grant_t = await wait_for_full_grant(fed, t, grant_timeout)
            model.observe_grant(grant_t)
            if model.output_handler():
                raise RuntimeError("time-advance model unexpectedly emitted output")
            model.internal_transition()

        print(
            f"regulator_main[{name}]: done — {len(model.grants)} grants",
            file=sys.stderr,
            flush=True,
        )

    return {
        "name": name,
        "lookahead": lookahead,
        "primitive": "NER",
        "constrained": constrained,
        "requested": model.requested,
        "grants": model.grants,
        "model_cycles": model.completed_cycles,
        "output_calls": model.output_calls,
        "ticks_sent": len(model.grants),
    }


def main(argv: list[str] | None = None) -> int:
    parser = common_parser("regulator_main")
    args = parser.parse_args(argv)
    try:
        result = asyncio.run(
            run(
                url=args.url,
                name=args.name,
                lookahead=args.lookahead,
                cycles=args.cycles,
                tick_step=args.tick_step,
                constrained=args.constrained,
                grant_timeout=args.grant_timeout,
            )
        )
    except Exception as exc:  # noqa: BLE001
        print(f"regulator_main[{args.name}]: {exc}", file=sys.stderr)
        return 1
    write_result(args.result, result)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/go-timed/regulator_main.go
// regulator_main.go — federate subprocess entry. Dials rtid via
// rti/pkg/federate, enables time regulation + (optionally) constrained,
// then runs `cycles` advance-request cycles. Each cycle issues either
// NER or TAR (per --primitive flag) and waits for the matching grant
// on the federate's Events() channel before issuing the next.
//
// On exit writes the result JSON: federate name, lookahead, primitive,
// list of grant times, and Tick send count.

package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/cbchoi/gorti/rti/pkg/federate"
)

func main() {
	if err := mainErr(); err != nil {
		fmt.Fprintln(os.Stderr, "regulator_main:", err)
		os.Exit(1)
	}
}

type mainArgs struct {
	url         string
	federation  string
	name        string
	lookahead   float64
	primitive   string // "NER" or "TAR"
	constrained bool
	cycles      int
	tickStep    float64 // logical time advance per cycle
	resultPath  string
	fomPath     string
}

func mainErr() error {
	args := parseArgs()

	// Read FOM bytes.
	xml, err := os.ReadFile(args.fomPath)
	if err != nil {
		return fmt.Errorf("read FOM: %w", err)
	}

	ctx := context.Background()
	conn, err := federate.Connect(ctx, args.url)
	if err != nil {
		return fmt.Errorf("connect %s: %w", args.url, err)
	}
	defer conn.Close()

	fmt.Fprintf(os.Stderr, "regulator_main[%s]: connected to %s\n", args.name, args.url)

	fed, err := conn.JoinFederation(ctx, federate.FederationSpec{
		Name:                args.federation,
		FOMModules:          []federate.FOMModule{{Path: args.fomPath, XML: xml}},
		StallTimeoutSeconds: 30,
	}, args.name)
	if err != nil {
		return fmt.Errorf("join: %w", err)
	}
	defer fed.Resign(ctx)

	fmt.Fprintf(os.Stderr, "regulator_main[%s]: joined as handle %d\n", args.name, fed.Handle())

	// Declare pub/sub Tick.
	if err := fed.PublishInteractionClass(ctx, "Tick"); err != nil {
		return fmt.Errorf("publish Tick: %w", err)
	}
	if err := fed.SubscribeInteractionClass(ctx, "Tick"); err != nil {
		return fmt.Errorf("subscribe Tick: %w", err)
	}

	// Enable regulation + (optionally) constrained.
	if err := fed.EnableTimeRegulation(ctx, args.lookahead); err != nil {
		return fmt.Errorf("EnableTimeRegulation: %w", err)
	}
	if args.constrained {
		if err := fed.EnableTimeConstrained(ctx); err != nil {
			return fmt.Errorf("EnableTimeConstrained: %w", err)
		}
	}

	r := &regulator{name: args.name, lookahead: args.lookahead}

	// Cycle loop. Each iteration issues an advance primitive, waits
	// for a *full* grant (M22 W3 — see waitForFullGrant for why
	// forced grants don't end the cycle), then emits a Tick at the
	// earliest valid TSO time after the grant.
	for i := 1; i <= args.cycles; i++ {
		t := float64(i) * args.tickStep
		switch args.primitive {
		case "NER":
			err = fed.NextMessageRequest(ctx, t)
		case "TAR":
			err = fed.TimeAdvanceRequest(ctx, t)
		default:
			return fmt.Errorf("unknown --primitive %q", args.primitive)
		}
		if err != nil {
			return fmt.Errorf("cycle %d %s(%v): %w", i, args.primitive, t, err)
		}

		// Wait for full grant. NER may produce intermediate forced
		// grants when this federate is sole-pending and LBTS < t;
		// per IEEE 1516.1 those are advisory ("messages with ts <=
		// LBTS are deliverable now") and the federate stays in
		// time-advancing-state until the full grant arrives. TAR
		// produces a single grant per request and finishes the cycle.
		grantT, err := waitForFullGrant(fed, t, args.primitive, 30*time.Second)
		if err != nil {
			return fmt.Errorf("cycle %d wait grant: %w", i, err)
		}
		r.grants = append(r.grants, grantT)

		// Emit a Tick at the earliest TSO time allowed after the grant.
		payload := r.nextTickPayload()
		sendT := r.tickTimestamp(grantT)
		if err := fed.SendInteraction(ctx, "Tick",
			map[string][]byte{"seq": payload}, &sendT); err != nil {
			return fmt.Errorf("cycle %d send Tick: %w", i, err)
		}
	}

	fmt.Fprintf(os.Stderr, "regulator_main[%s]: done — %d grants\n", args.name, len(r.grants))

	return writeResult(args.resultPath, r, args)
}

// waitForFullGrant drains f.Events() until a *full* TimeAdvanceGrant
// is observed (grant.Time >= requested), accumulating any
// intermediate forced grants (NER/NMRA only — TAR/TARA/FQR clear
// pending on every grant per advance.go::decideGrant). For TAR the
// loop returns on the first grant since one-grant-per-request is
// TAR's contract.
//
// M22 W3: this replaces the M21-era waitForGrant + 5ms settle delay.
// The settle delay was masking the SDK-side semantic gap, not a
// server race. Per spec a federate stays in time-advancing-state
// across forced grants; reissuing an advance primitive at that point
// correctly returns ErrTimeAdvancingState.
func waitForFullGrant(f *federate.Federate, requested float64, primitive string, timeout time.Duration) (float64, error) {
	tarLike := primitive == "TAR" || primitive == "TARA" || primitive == "FQR"
	deadline := time.NewTimer(timeout)
	defer deadline.Stop()
	for {
		select {
		case ev, ok := <-f.Events():
			if !ok {
				return 0, fmt.Errorf("events channel closed before grant")
			}
			g, ok := ev.(federate.TimeAdvanceGrant)
			if !ok {
				// ReceiveInteraction etc.: ignore.
				continue
			}
			if tarLike || g.Time >= requested {
				return g.Time, nil
			}
			// NER/NMRA forced grant (g.Time < requested) — accumulate
			// and keep waiting.
			fmt.Fprintf(os.Stderr, "regulator: forced grant @ %v < requested %v; waiting for full\n", g.Time, requested)
		case <-deadline.C:
			return 0, fmt.Errorf("deadline waiting for full grant (requested=%v)", requested)
		}
	}
}

func writeResult(path string, r *regulator, args mainArgs) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	payload := map[string]any{
		"name":        r.name,
		"lookahead":   r.lookahead,
		"primitive":   args.primitive,
		"constrained": args.constrained,
		"grants":      r.grants,
		"ticks_sent":  r.tickSeq,
	}
	b, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, b, 0o644)
}

func parseArgs() mainArgs {
	var a mainArgs
	flag.StringVar(&a.url, "url", "127.0.0.1:8442", "rtid host:port")
	flag.StringVar(&a.federation, "federation", "go-timed", "federation name")
	flag.StringVar(&a.name, "name", "fast", "federate name (fast|normal|slow)")
	flag.Float64Var(&a.lookahead, "lookahead", 0.5, "regulation lookahead")
	flag.StringVar(&a.primitive, "primitive", "NER", "advance primitive (NER|TAR)")
	flag.BoolVar(&a.constrained, "constrained", true, "enable time-constrained")
	flag.IntVar(&a.cycles, "cycles", 10, "number of advance cycles")
	flag.Float64Var(&a.tickStep, "tick-step", 1.0, "logical time advance per cycle")
	flag.StringVar(&a.resultPath, "result", "/tmp/go-timed/result.json", "result JSON path")
	flag.StringVar(&a.fomPath, "fom", "examples/go-timed/time-advance-fom.xml", "FOM XML path")
	flag.Parse()
	return a
}
